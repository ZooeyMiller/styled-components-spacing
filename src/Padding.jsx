import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {map} from 'styled-components-breakpoint';
import sizes from './sizes';

const getSizeFromTheme = (theme, size) => theme && theme.spacing && theme.spacing[size] || sizes[size]; //TODO: in dev, throw an error if size doesn't exist
const getBreakpointsFromTheme = theme => theme && theme.breakpoints;

export const p = (size, theme) => map(size, s => `
  padding: ${getSizeFromTheme(theme, s)};
`, getBreakpointsFromTheme(theme));

export const px = (size, theme) => map(size, s => `
  padding-left: ${getSizeFromTheme(theme, s)};
  padding-right: ${getSizeFromTheme(theme, s)};
`, getBreakpointsFromTheme(theme));

export const py = (size, theme) => map(size, s => `
  padding-top: ${getSizeFromTheme(theme, s)};
  padding-bottom: ${getSizeFromTheme(theme, s)};
`, getBreakpointsFromTheme(theme));

export const pt = (size, theme) => map(size, s => `
  padding-top: ${getSizeFromTheme(theme, s)};
`, getBreakpointsFromTheme(theme));

export const pr = (size, theme) => map(size, s => `
  padding-right: ${getSizeFromTheme(theme, s)};
`, getBreakpointsFromTheme(theme));

export const pb = (size, theme) => map(size, s => `
  padding-bottom: ${getSizeFromTheme(theme, s)};
`, getBreakpointsFromTheme(theme));

export const pl = (size, theme) => map(size, s => `
  padding-left: ${getSizeFromTheme(theme, s)};
`, getBreakpointsFromTheme(theme));

export const Padding = styled.div`
  ${({inline}) => inline && 'display: inline-block;'}
  ${({all, theme}) => p(all, theme)}
  ${({horizontal, theme}) => px(horizontal, theme)} 
  ${({vertical, theme}) => py(vertical, theme)} 
  ${({top, theme}) => pt(top, theme)} 
  ${({right, theme}) => pr(right, theme)} 
  ${({bottom, theme}) => pb(bottom, theme)} 
  ${({left, theme}) => pl(left, theme)} 
`;

Padding.propTypes = {
  all: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.object]),
  horizontal: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.object]),
  vertical: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.object]),
  top: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.object]),
  right: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.object]),
  bottom: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.object]),
  left: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.object]),
  inline: PropTypes.bool
};

export default Padding;
