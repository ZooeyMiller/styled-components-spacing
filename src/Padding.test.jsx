import 'jest-styled-components';
import React from 'react';
import renderer from 'react-test-renderer';
import { Padding } from './Padding';

describe('Padding', () => {
  it('it should have no styles when no props are passed', () => {
    const tree = renderer.create(<Padding />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('inline', () => {
    it('it should be set when the prop is true', () => {
      const tree = renderer.create(<Padding inline={true} />).toJSON();
      expect(tree).toMatchSnapshot();
      expect(tree).toHaveStyleRule('display', 'inline-block');
    });

    it('it should not be set when the prop is false', () => {
      const tree = renderer.create(<Padding inline={false} />).toJSON();
      expect(tree).not.toHaveStyleRule('display', 'inline-block');
    });
  });

  describe('all', () => {
    it('it should have a single padding for all breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding all={1} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('it should have a different padding for all breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding all={{ mobile: 0, tablet: 2, desktop: 4 }} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('it should have a different padding for some breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding all={{ tablet: 2, desktop: 4 }} />).toJSON();
      expect(tree).toMatchSnapshot();
      expect(tree).toHaveStyleRule('padding', '0.5rem', {
        media: '(min-width:46.0625em)'
      });
      expect(tree).toHaveStyleRule('padding', '2rem', {
        media: '(min-width:64.0625em)'
      });
    });
  });

  describe('horizontal', () => {
    it('it should have a single padding for all breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding horizontal={1} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('it should have a different padding for all breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding horizontal={{ mobile: 0, tablet: 2, desktop: 4 }} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('it should have a different padding for some breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding horizontal={{ tablet: 2, desktop: 4 }} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('vertical', () => {
    it('it should have a single padding for all breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding vertical={1} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('it should have a different padding for all breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding vertical={{ mobile: 0, tablet: 2, desktop: 4 }} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('it should have a different padding for some breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding vertical={{ tablet: 2, desktop: 4 }} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('top', () => {
    it('it should have a single padding for all breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding top={1} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('it should have a different padding for all breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding top={{ mobile: 0, tablet: 2, desktop: 4 }} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('it should have a different padding for some breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding top={{ tablet: 2, desktop: 4 }} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('right', () => {
    it('it should have a single padding for all breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding right={1} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('it should have a different padding for all breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding right={{ mobile: 0, tablet: 2, desktop: 4 }} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('it should have a different padding for some breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding right={{ tablet: 2, desktop: 4 }} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('bottom', () => {
    it('it should have a single padding for all breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding bottom={1} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('it should have a different padding for all breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding bottom={{ mobile: 0, tablet: 2, desktop: 4 }} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('it should have a different padding for some breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding bottom={{ tablet: 2, desktop: 4 }} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('left', () => {
    it('it should have a single padding for all breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding left={1} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('it should have a different padding for all breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding left={{ mobile: 0, tablet: 2, desktop: 4 }} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('it should have a different padding for some breakpoints when the prop is passed', () => {
      const tree = renderer.create(<Padding left={{ tablet: 2, desktop: 4 }} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
