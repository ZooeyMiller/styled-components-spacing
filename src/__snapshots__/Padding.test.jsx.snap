// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Padding all it should have a different padding for all breakpoints when the prop is passed 1`] = `
.c0 {
  padding: 0;
}

@media (min-width:46.0625em) {
  .c0 {
    padding: 0.5rem;
  }
}

@media (min-width:64.0625em) {
  .c0 {
    padding: 2rem;
  }
}

<div
  className="c0"
/>
`;

exports[`Padding all it should have a different padding for some breakpoints when the prop is passed 1`] = `
@media (min-width:46.0625em) {
  .c0 {
    padding: 0.5rem;
  }
}

@media (min-width:64.0625em) {
  .c0 {
    padding: 2rem;
  }
}

<div
  className="c0"
/>
`;

exports[`Padding all it should have a single padding for all breakpoints when the prop is passed 1`] = `
.c0 {
  padding: 0.25rem;
}

<div
  className="c0"
/>
`;

exports[`Padding bottom it should have a different padding for all breakpoints when the prop is passed 1`] = `
.c0 {
  padding-bottom: 0;
}

@media (min-width:46.0625em) {
  .c0 {
    padding-bottom: 0.5rem;
  }
}

@media (min-width:64.0625em) {
  .c0 {
    padding-bottom: 2rem;
  }
}

<div
  className="c0"
/>
`;

exports[`Padding bottom it should have a different padding for some breakpoints when the prop is passed 1`] = `
@media (min-width:46.0625em) {
  .c0 {
    padding-bottom: 0.5rem;
  }
}

@media (min-width:64.0625em) {
  .c0 {
    padding-bottom: 2rem;
  }
}

<div
  className="c0"
/>
`;

exports[`Padding bottom it should have a single padding for all breakpoints when the prop is passed 1`] = `
.c0 {
  padding-bottom: 0.25rem;
}

<div
  className="c0"
/>
`;

exports[`Padding horizontal it should have a different padding for all breakpoints when the prop is passed 1`] = `
.c0 {
  padding-left: 0;
  padding-right: 0;
}

@media (min-width:46.0625em) {
  .c0 {
    padding-left: 0.5rem;
    padding-right: 0.5rem;
  }
}

@media (min-width:64.0625em) {
  .c0 {
    padding-left: 2rem;
    padding-right: 2rem;
  }
}

<div
  className="c0"
/>
`;

exports[`Padding horizontal it should have a different padding for some breakpoints when the prop is passed 1`] = `
@media (min-width:46.0625em) {
  .c0 {
    padding-left: 0.5rem;
    padding-right: 0.5rem;
  }
}

@media (min-width:64.0625em) {
  .c0 {
    padding-left: 2rem;
    padding-right: 2rem;
  }
}

<div
  className="c0"
/>
`;

exports[`Padding horizontal it should have a single padding for all breakpoints when the prop is passed 1`] = `
.c0 {
  padding-left: 0.25rem;
  padding-right: 0.25rem;
}

<div
  className="c0"
/>
`;

exports[`Padding inline it should be set when the prop is true 1`] = `
.c0 {
  display: inline-block;
}

<div
  className="c0"
/>
`;

exports[`Padding it should have no styles when no props are passed 1`] = `
<div
  className=""
/>
`;

exports[`Padding left it should have a different padding for all breakpoints when the prop is passed 1`] = `
.c0 {
  padding-left: 0;
}

@media (min-width:46.0625em) {
  .c0 {
    padding-left: 0.5rem;
  }
}

@media (min-width:64.0625em) {
  .c0 {
    padding-left: 2rem;
  }
}

<div
  className="c0"
/>
`;

exports[`Padding left it should have a different padding for some breakpoints when the prop is passed 1`] = `
@media (min-width:46.0625em) {
  .c0 {
    padding-left: 0.5rem;
  }
}

@media (min-width:64.0625em) {
  .c0 {
    padding-left: 2rem;
  }
}

<div
  className="c0"
/>
`;

exports[`Padding left it should have a single padding for all breakpoints when the prop is passed 1`] = `
.c0 {
  padding-left: 0.25rem;
}

<div
  className="c0"
/>
`;

exports[`Padding right it should have a different padding for all breakpoints when the prop is passed 1`] = `
.c0 {
  padding-right: 0;
}

@media (min-width:46.0625em) {
  .c0 {
    padding-right: 0.5rem;
  }
}

@media (min-width:64.0625em) {
  .c0 {
    padding-right: 2rem;
  }
}

<div
  className="c0"
/>
`;

exports[`Padding right it should have a different padding for some breakpoints when the prop is passed 1`] = `
@media (min-width:46.0625em) {
  .c0 {
    padding-right: 0.5rem;
  }
}

@media (min-width:64.0625em) {
  .c0 {
    padding-right: 2rem;
  }
}

<div
  className="c0"
/>
`;

exports[`Padding right it should have a single padding for all breakpoints when the prop is passed 1`] = `
.c0 {
  padding-right: 0.25rem;
}

<div
  className="c0"
/>
`;

exports[`Padding top it should have a different padding for all breakpoints when the prop is passed 1`] = `
.c0 {
  padding-top: 0;
}

@media (min-width:46.0625em) {
  .c0 {
    padding-top: 0.5rem;
  }
}

@media (min-width:64.0625em) {
  .c0 {
    padding-top: 2rem;
  }
}

<div
  className="c0"
/>
`;

exports[`Padding top it should have a different padding for some breakpoints when the prop is passed 1`] = `
@media (min-width:46.0625em) {
  .c0 {
    padding-top: 0.5rem;
  }
}

@media (min-width:64.0625em) {
  .c0 {
    padding-top: 2rem;
  }
}

<div
  className="c0"
/>
`;

exports[`Padding top it should have a single padding for all breakpoints when the prop is passed 1`] = `
.c0 {
  padding-top: 0.25rem;
}

<div
  className="c0"
/>
`;

exports[`Padding vertical it should have a different padding for all breakpoints when the prop is passed 1`] = `
.c0 {
  padding-top: 0;
  padding-bottom: 0;
}

@media (min-width:46.0625em) {
  .c0 {
    padding-top: 0.5rem;
    padding-bottom: 0.5rem;
  }
}

@media (min-width:64.0625em) {
  .c0 {
    padding-top: 2rem;
    padding-bottom: 2rem;
  }
}

<div
  className="c0"
/>
`;

exports[`Padding vertical it should have a different padding for some breakpoints when the prop is passed 1`] = `
@media (min-width:46.0625em) {
  .c0 {
    padding-top: 0.5rem;
    padding-bottom: 0.5rem;
  }
}

@media (min-width:64.0625em) {
  .c0 {
    padding-top: 2rem;
    padding-bottom: 2rem;
  }
}

<div
  className="c0"
/>
`;

exports[`Padding vertical it should have a single padding for all breakpoints when the prop is passed 1`] = `
.c0 {
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
}

<div
  className="c0"
/>
`;
